go-controller/pkg/node/rulemanager/rule_manager.go

```diff
@@ -0,0 +1,182 @@
+	rm.mu.Lock()
+	defer rm.mu.Unlock()
+	// check if we are already managing this route and if so, no-op
+	for _, existingRule := range rm.rules {
+		if areNetlinkRulesEqual(existingRule.rule, &rule) {
+			return nil
+		}
+	}
+	rm.rules = append(rm.rules, nlRule{rule: &rule})
+	return rm.reconcile()
+}
+
+// DeleteRule stops managed a rule and ensures its deleted
+func (rm *Controller) DeleteRule(rule netlink.Rule) error {
+	rm.mu.Lock()
+	defer rm.mu.Unlock()
+	var reconcileNeeded bool
+	for i, r := range rm.rules {
+		if areNetlinkRulesEqual(r.rule, &rule) {
+			rm.rules[i].delete = true
+			reconcileNeeded = true
+			break
+		}
+	}
+	if reconcileNeeded {
+		return rm.reconcile()
+	}
+	return nil
+}
+
+// OwnPriority ensures any rules observed with priority 'priority' must be specified otherwise its removed
+func (rm *Controller) OwnPriority(priority int) error {
+	rm.mu.Lock()
+	defer rm.mu.Unlock()
+	rm.ownPriorities[priority] = true
+	return rm.reconcile()
+}
+
+func (rm *Controller) reconcile() error {
+	start := time.Now()
+	defer func() {
+		klog.V(5).Infof("Reconciling rules took %v", time.Since(start))
+	}()
+	var family int
+	if rm.v4 && rm.v6 {
+		family = netlink.FAMILY_ALL
+	} else if rm.v4 {
+		family = netlink.FAMILY_V4
+	} else if rm.v6 {
+		family = netlink.FAMILY_V6
+	}
+
+	rulesFound, err := netlink.RuleList(family)
+	if err != nil {
+		return err
+	}
+	var errors []error
+	rulesToKeep := make([]nlRule, 0)
+	for _, r := range rm.rules {
+		// delete rule by first checking if it exists and if so, delete it
+		if r.delete {
+			if found, foundRoute := isNetlinkRuleInSlice(rulesFound, r.rule); found {
+				if err = netlink.RuleDel(foundRoute); err != nil {
+					// retry later
+					rulesToKeep = append(rulesToKeep, r)
+					errors = append(errors, err)
+				}
+			}
+		} else {
+			// add rule by first checking if it exists and if not, add it
+			rulesToKeep = append(rulesToKeep, r)
+			if found, _ := isNetlinkRuleInSlice(rulesFound, r.rule); !found {
+				if err = netlink.RuleAdd(r.rule); err != nil {
+					errors = append(errors, err)
+				}
+			}
+		}
+	}
+
+	var found bool
+	for priority, _ := range rm.ownPriorities {
+		for _, ruleFound := range rulesFound {
+			if ruleFound.Priority != priority {
+				continue
+			}
+			found = false
+			for _, ruleWanted := range rm.rules {
+				if ruleWanted.rule.Priority != priority {
+					continue
+				}
+				if areNetlinkRulesEqual(ruleWanted.rule, &ruleFound) {
+					found = true
+					break
+				}
+			}
+			if !found {
+				klog.Infof("Rule manager: deleting stale rule (%s) found at priority %d", ruleFound.String(), priority)
+				if err = netlink.RuleDel(&ruleFound); err != nil {
+					klog.Errorf("Rule manager: failed to delete stale rule (%s) found at priority %d", ruleFound.String(), priority)
+				}
+			}
+		}
+	}
+
+	rm.rules = rulesToKeep
+	return utilerrors.NewAggregate(errors)
+}
+
+func areNetlinkRulesEqual(r1, r2 *netlink.Rule) bool {
```

Comment:
why is priority not compared in this function?

Reply:
it is being compared - here - here is the rule struct:
```
// Rule represents a netlink rule.
type Rule struct {
	Priority          int
	Family            int
	Table             int
	Mark              int
	Mask              int
	Tos               uint
	TunID             uint
	Goto              int
	Src               *net.IPNet
	Dst               *net.IPNet
	Flow              int
	IifName           string
	OifName           string
	SuppressIfgroup   int
	SuppressPrefixlen int
	Invert            bool
	Dport             *RulePortRange
	Sport             *RulePortRange
	IPProto           int
	UIDRange          *RuleUIDRange
	Protocol          uint8
}
```

Here is the .String() func for this type:

```
func (r Rule) String() string {
	from := "all"
	if r.Src != nil && r.Src.String() != "<nil>" {
		from = r.Src.String()
	}

	to := "all"
	if r.Dst != nil && r.Dst.String() != "<nil>" {
		to = r.Dst.String()
	}

	return fmt.Sprintf("ip rule %d: from %s to %s table %d",
		r.Priority, from, to, r.Table)
}
```